keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,969 @@
+3. New policy selectors (services, service accounts, etc.)
+4. Tools/CLI to debug/explain NetworkPolicies, cluster-scoped policies and their
+   impact on workloads.
+
+## Proposal
+
+In order to achieve the two primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following two new resources
+under `networking.k8s.io` API group:
+- ClusterNetworkPolicy
+- DefaultNetworkPolicy
+
+### ClusterNetworkPolicy resource
+
+A ClusterNetworkPolicy resource will help the administrators set strict
+security rules for the cluster, i.e. a developer CANNOT override these rules
+by creating NetworkPolicies that applies to the same workloads as the
+ClusterNetworkPolicy does.
+
+Unlike the NetworkPolicy resource in which each rule represents a allowed
+traffic, ClusterNetworkPolicy will enable administrators to set `Allow` or
+`Deny` as the action of each rule. ClusterNetworkPolicy rules should be read
+as-is, i.e. there will not be any implicit isolation effects for the Pods
+selected by the ClusterNetworkPolicy, as opposed to what NetworkPolicy rules imply.
+
+In terms of precedence, the aggregated `Deny` rules (all ClusterNetworkPolicy
+rules with action `Deny` in the cluster combined) should be evaluated before
+aggregated ClusterNetworkPolicy `Allow` rules, followed by aggregated
+NetworkPolicy rules in all Namespaces. As such, the `Deny` rules have the
+highest precedence, which prevents them to be unexpectedly overwritten.
+
+ClusterNetworkPolicy `Deny` rules are useful for administrators to explicitly
+block traffic from malicious clients, or workloads that poses security risks.
+Those traffic restrictions can only be lifted once the`Deny` rules are deleted
+or modified. On the other hand, the `Allow` rules can be used to call out
+traffic in the cluster that needs to be allowed for certain components to work
+as expected (egress to CoreDNS for example). Those traffic should not be blocked
+when developers apply NetworkPolicy to their Namespaces which isolates the workloads.
+
+### DefaultNetworkPolicy resource
+
+A DefaultNetworkPolicy resource will help the administrators set baseline security
+rules for the cluster, i.e. a developer CAN override these rules by creating
+NetworkPolicies that applies to the same workloads as the DefaultNetworkPolicy does.
+
+DefaultNetworkPolicy works just like NetworkPolicy except that it is cluster-scoped.
+When workloads are selected by a DefaultNetworkPolicy, they are isolated except
+for the ingress/egress rules specified. DefaultNetworkPolicy rules will not have
+actions associated -- each rule will be an 'allow' rule.
+
+Aggregated NetworkPolicy rules will be evaluated before aggregated DefaultNetworkPolicy rules.
+If a Pod is selected by both, a DefaultNetworkPolicy and a NetworkPolicy, then
+the DefaultNetworkPolicy's effect on that Pod becomes obsolete.
+In this case, the traffic allowed will be solely determined by the NetworkPolicy.
+
+### Precedence model
```

Comment:
DNP is also supposed to allow for overriding the baseline accept/deny policy. That is, kubernetes clusters by default have the rule "pods that are selected by no NetworkPolicies allow everything", but with DNP you can change that to "pods that are selected by no NetworkPolicies deny everything" or even "pods that are selected by no NetworkPolicies allow from other pods in the same namespace but not from pods in other namespaces". You can't implement that with your suggestion.

(The counterargument would be that this is really a totally different kind of thing from "I want to block pods from talking to A.B.C.D" / "I want to force pods to accept connections from `monitoring-namespace`", and should be defined as a separate object, rather than trying to squish both pieces of functionality together.)

Reply:
> DNP is also supposed to allow for overriding the baseline accept/deny policy. 
Thanks for calling it out @danwinship. To your counter-argument, we also thought about making this baseline intent a config option for the cluster for example, but it will be hard for admins to fine-tune it and dynamically control the baseline postures.

Reply:
we incorporated @tgraf suggestion of `exceptions` (named `authorize`).


for the `default` action, there was quite a bit of push back when it was first proposed in sig-network, hence keeping it as is in the form of 2 CRDs for now

Reply:
Yeah that was my feeling about DNP as well and I believe I mentioned it somewhere else. I think the reasoning is to avoid needing to rubber-stamp resources into each namespace. But like you said I'm not sure if that's justification for a new resource. It can just be scripted with a controller or something.

Reply:
> As it is defined (user NetPol totally deactivates it), how is it different than just putting a NetworkPolicy resource into the user's namespace?

Because you can write a DNP that sets the baseline policy for all namespaces, including namespaces that haven't yet been created, but there's no way to ensure that a NetworkPolicy gets automatically created in every newly-created namespace. And there's a big difference between "all Namespaces are isolated by default" and "please remember to make all of your Namespaces isolated after you create them!"

Additionally, this preserves the distinction between administrator intent and developer/deployer intent; NetworkPolicies are for developers to use, CNPs and DNPs are for administrators to use.

Reply:
The namespace/ tenant isolation requirement can be either admin-controlled or user-controlled in practice:  Either the (a) cluster admin knows and enforces the grouping of "co-operating namespaces" which make up a tenant or (b) the cluster admin lets the users/ tenants fully control whichever set of namespaces they want to group/ allow communication and has no direct visibility or control.   For (a), the namespace group isolation should be enforced via CNP and not via DNP. For (b) If the cluster admin can only inject initial config but has no easy way of verifying what parts of the DNP are being over-riden by the namespace NPs, then the cluster admin probably wont care to use the DNP as a tool. For a true self-service namespace grouping/ multitenancy model, there need to be other ways to let the cluster admin have visibility and predictability while retaining self-service ability for the end users.

It seems the combination of an admin-generated DNP + user generated NP has unpredictable behavior for the cluster admin so they wont care to use it very much. So either not have the DNP at all or decide that once a user provisions an NP, then all DNPs in that namespace become dont cares).

Also I would think that app developers would want to write cluster independent network policies where possible to ensure portability hence cant assume that DNP in cluster A would be the same as DNP in cluster B so would anyway cover all their requirements via NPs rather than relying on the cluster-dependent combination of NP + DNP.

<ended up being a longer comment than I'd have liked :-)>

Reply:
Also we may want to revisit and consider having simple priority ordering (sequence numbers etc) instead of various combinations of Authorize, Except, AlwaysAllow, Default etc with implicit priorities based on action type.  Also note that we would want to be able to add future actions like "mirror packet" or "log packet" and priority of the rule should not be overloaded and hard coded into the action type.

Reply:
> Either the (a) cluster admin knows and enforces the grouping of "co-operating namespaces" which make up a tenant or (b) the cluster admin lets the users/ tenants fully control whichever set of namespaces they want to group/ allow communication and has no direct visibility or control. For (a), the namespace group isolation should be enforced via CNP and not via DNP. For (b) If the cluster admin can only inject initial config but has no easy way of verifying what parts of the DNP are being over-riden by the namespace NPs, then the cluster admin probably wont care to use the DNP as a tool.

There are admins who want "I want to lock down everything my way and not let users bypass it" and they can use CNP. But there are also lots of admins who want "I want the default to be locked-down, so there's reasonable baseline security, and developers have to explicitly specify what sort of ingress/egress access they expect".

Reply:
^^ Brief update from thoughts in the background while we were discussing this KEP today, even though we did not specifically discuss these proposals. Option (1) above is not explicit enough, and if it introduces a "default deny" then it begs the question how the policy applies to cluster-external traffic. We are not aiming to address cluster-external traffic in this KEP, and I don't think that it would be reasonable to define implicit behaviour on the policy impact for cluster-internal traffic only. I think it's pretty safe to discard Option (1) above.

Reply:
since we moved on to a newer proposal including priorities we can resolve this comment as it no longer applies